  
## Generatiivinen tekstimallin testaus ja analyysi

In [1]:

import keras
import sentencepiece as spm

model = keras.models.load_model('satuteksti_model.keras')
sp = spm.SentencePieceProcessor()
sp.load('satuteksti_sp.model')

True

### Analysoida luodun tekstin laatua

#### käyttää Pyvoikko:

- kielioppivirheiden tarkistamiseen.

- Textstat luettavuuden ja monimutkaisuuden arvioimiseen.

    - Sanaston monipuolisuuteen esim. tyypillisten sanojen ja uniikkien sanojen määrää.
- Luettavuuspiste tarkistamiseen.

In [2]:
import pyvoikko
import textstat
from saved_functions.generate_text_funktio import generate_text

prompts = [
    "prinsessa ja leipuri",
    "sinä ja minä",
    "Lintu laulaa"
]

analyse_num = 1 # Generate texts from prompts
for prompt in prompts:
    generated_text = generate_text(model, sp, prompt, num_tokens=100, temperature=1.0)
    print(f"Transcribed text: {generated_text}")

    words = generated_text.split()
    unique_words = set(words)
    vocab_diversity = len(unique_words) / len(words) if words else 0
    readability_score = textstat.flesch_reading_ease(generated_text)

    print(f"\n Analyysin yhteenveto: {analyse_num}")
    analyse_num += 1
    error_count = 0

    for word in words:
        if not pyvoikko.analyse(word):
            error_count += 1

    print(f"- Virheellisiä sanoja: {error_count}")
    print(f"- Sanaston monipuolisuus (uniikit / kaikki): {vocab_diversity:.2f}")
    print(f"- Luettavuuspiste (Flesch): {readability_score:.2f}")
    print("")

Transcribed text: prinsessa ja leipuri poistui verkalleen. Vitkaan astui hän alas linnan marmoriportaita. — Parasta lienee laittaa kuntoon kotinsa ja talonsa ja valmistautua eroamaan täältä, pitääkseen kärpäsiä loitolla. Leipuri oli sitoi kiinni satulavyöllä ja leveämpi kuin naapuriprinssi. Mutta ratkaisevana hetkenä ei tämäkään prinssi saanut pitkin. Mutta kun hän ei hylännyt, ja mihin tuli hänen oli ajanut kotiin päin, niin että ajoneuvot. Miehet kiskoivat ohjaksista ja löivät hevoskoneja. — Aha, sanoi Niilo, eipä minulle, teillä vai minulla, näet porvarit kaupungissa, sekä talonpojat maalla toivoivat prinssin pian ottavan itselleen puolison. He

 Analyysin yhteenveto: 1
- Virheellisiä sanoja: 25
- Sanaston monipuolisuus (uniikit / kaikki): 0.89
- Luettavuuspiste (Flesch): 52.26

Transcribed text: sinä ja minä pääsen eteenpäin. — Et sinä niin vähällä pääse! sanoi haltijatar. — Saammepahan nähdä, kenellä meistä kärsivällisyyttä enemmän riittää, sinulla vai minulla, sanoi Niilo, myllyr

### Analyysin yhteenveto:

Alla olevassa taulukossa esitetään kolmen analyysin tulokset. Taulukosta näkee, kuinka monta virheellistä sanaa on kussakin analyysissä, sanaston monipuolisuuden (eli uniikkien sanojen suhde kaikkiin sanoihin) sekä tekstin luettavuuden Fleschin kaavan mukaan arvioituna.

| **Analyysi** | **Virheellisiä sanoja** | **Sanaston monipuolisuus** | **Luettavuuspiste (Flesch)** |
|--------------|-------------------------|----------------------------|------------------------------|
| **Analyysi 1** | 25                      | 0.89                       | 52.26                        |
| **Analyysi 2** | 28                      | 0.88                       | 43.09                        |
| **Analyysi 3** | 29                      | 0.82                       | 66.23                        |

**Huomioita:**
- **Analyysi 1** osoittaa vahvaa sanaston monipuolisuutta ja hyvää luettavuutta verrattuna muihin.
- **Analyysi 2** sisältää hieman enemmän virheitä ja matalamman luettavuuspisteen.
- **Analyysi 3** on heikoin sanaston monipuolisuudessa, mutta sai korkeimman luettavuuspisteen, mikä voi viitata yksinkertaisempaan kieleen.
